File: RGB_Control/Form1.cs
using System;
using System.Drawing;
using System.IO.Ports;
using System.Net;
using System.Net.Sockets;
using System.Windows.Forms;
using Arduino_LED;

namespace RGB_Control
{
    public partial class Form1 : Form
    {
        private Byte[] cmd = { 0, 0, 0, 255 };
        private TcpClient client;
        private NetworkStream steam;
        private MyRGB color;
        
        public Form1()
        {
            InitializeComponent();
            serialPort1.ReadTimeout = 3000;          
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            color = new MyRGB();
        }

        /////////////// Control Group /////////////////
        /// Panel
        private void panel1_Click(object sender, EventArgs e)
        {
            DialogResult colorResult = colorDialog1.ShowDialog();
            if(colorResult == DialogResult.OK)
            {
                panel1.BackColor = colorDialog1.Color;
                tbRed.Value = colorDialog1.Color.R;
                tbGreen.Value = colorDialog1.Color.G;
                tbBlue.Value = colorDialog1.Color.B;
                color.setRGB(colorDialog1.Color.R, colorDialog1.Color.G, colorDialog1.Color.B);
            }
        }

        private void btnUpdateColor_Click(object sender, EventArgs e)
        {
            if (serialPort1.IsOpen)
            {
                ControlMessage();
                serialPort1.Write(cmd,0,4);
            }
            else if (client.Connected)
            {
                ControlMessage();
                steam.Write(cmd, 0, 4);
            }
        }

        /// Track bar
        private void tbRed_ValueChanged(object sender, EventArgs e)
        {
            txbRed.Text = tbRed.Value.ToString();
            color.setRED(tbRed.Value);
        }

        private void tbGreen_ValueChanged(object sender, EventArgs e)
        {
            txbGreen.Text = tbGreen.Value.ToString();
            color.setGreen(tbGreen.Value);
        }

        private void tbBlue_ValueChanged(object sender, EventArgs e)
        {
            txbBlue.Text = tbBlue.Value.ToString();
            color.setBLUE(tbBlue.Value);
        }

        private void tbPower_ValueChanged(object sender, EventArgs e)
        {
            txbPower.Text = tbPower.Value.ToString();
            color.LEDPower = tbPower.Value;
        }

        private void tbRGB_Scroll(object sender, EventArgs e)
        {
            panel1.BackColor = Color.FromArgb(color.rgb.R, color.rgb.G, color.rgb.B);
        }
        
        /// Text box
        private void txbRed_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode.Equals(Keys.Return))
            {
                try
                {
                    int data = Int32.Parse(txbRed.Text);
                    if (data <= 255)
                    {
                        tbRed.Value = data;
                        color.setRED(data);
                        panel1.BackColor = Color.FromArgb(color.rgb.R, color.rgb.G, color.rgb.B);
                    }
                }
                catch (Exception)
                {

                }
            }
        }

        private void txbGreen_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode.Equals(Keys.Return))
            {
                try
                {
                    int data = Int32.Parse(txbGreen.Text);
                    if (data <= 255)
                    {
                        tbGreen.Value = data;
                        color.setGreen(data);
                        panel1.BackColor = Color.FromArgb(color.rgb.R, color.rgb.G, color.rgb.B);
                    }
                }
                catch (Exception)
                {

                }
            }
        }

        private void txbBlue_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode.Equals(Keys.Return))
            {
                try
                {
                    int data = Int32.Parse(txbBlue.Text);
                    if (data <= 255)
                    {
                        tbBlue.Value = data;
                        color.setBLUE(data);
                        panel1.BackColor = Color.FromArgb(color.rgb.R, color.rgb.G, color.rgb.B);
                    }
                }
                catch (Exception)
                {

                }
            }
        }

        private void txbPower_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode.Equals(Keys.Return))
            {
                try
                {
                    int data = Int32.Parse(txbPower.Text);
                    if (data <= 255)
                    {
                        tbPower.Value = data;
                        color.LEDPower = data;
                    }
                }
                catch (Exception)
                {

                }
            }
        }
        
       
        //////////////// Communication Group ///////////////
        /// Serial
        private void rdbSerial_Enter(object sender, EventArgs e)
        {
            gbSerial.Enabled = true;
            gbTCP.Enabled = false;
            cmbPortName.Enabled = true;
            cmbBaudRate.Enabled = true;
            btnSerialConnect.Enabled = false;
            btnSerialDiscon.Enabled = false;
            GetAvailablePort();
            try
            {
                client.Close();
                btnUpdateColor.Enabled = false;
            }
            catch (Exception) { }                  
        }

        private void SerialComSetupChange(object sender, EventArgs e)
        {
            try
            {
                if (cmbPortName.Text != "" && cmbBaudRate.Text != "")
                {
                    serialPort1.PortName = cmbPortName.Text;
                    serialPort1.BaudRate = Convert.ToInt32(cmbBaudRate.Text);
                    btnSerialConnect.Enabled = true;
                }
            }
            catch (UnauthorizedAccessException)
            {
                MessageBox.Show("Connection error");
            }
        }

        private void btnSerialConnect_Click(object sender, EventArgs e)
        {
            try
            {
                serialPort1.Open();
                btnSerialDiscon.Enabled = true;
                cmbPortName.Enabled = false;
                cmbBaudRate.Enabled = false;
                btnUpdateColor.Enabled = true;
            }
            catch (Exception)
            {
                MessageBox.Show("Connect error!!!");
                //throw;
            }
        }

        private void btnSerialDiscon_Click(object sender, EventArgs e)
        {
            try
            {
                serialPort1.Close();
                btnSerialConnect.Enabled = false;
                btnSerialDiscon.Enabled = false;
                cmbPortName.Enabled = true;
                cmbBaudRate.Enabled = true;
                btnUpdateColor.Enabled = false;
            }
            catch (Exception)
            {
                MessageBox.Show("Disconnect error!!!");
                throw;
            }
        }

        /// TCP
        private void rdbTCP_Enter(object sender, EventArgs e)
        {
            gbSerial.Enabled = false;
            gbTCP.Enabled = true;
            btnTCPConnect.Enabled = true;
            btnTCPdiscon.Enabled = false;
            if (serialPort1.IsOpen)
            {
                serialPort1.Close();
                btnUpdateColor.Enabled = false;
            }

        }

        private void btnTCPConnect_Click(object sender, EventArgs e)
        {
            TCPClientInitail();
            try
            {
                if (client.Connected)
                {
                    btnUpdateColor.Enabled = true;
                    btnTCPConnect.Enabled = false;
                    btnTCPdiscon.Enabled = true;
                }
            }
            catch (Exception)
            {
                MessageBox.Show("Connect time out!");
            }
            
            
        }

        private void btnTCPdiscon_Click(object sender, EventArgs e)
        {
            try
            {
                client.Close();
                btnUpdateColor.Enabled = false;
                btnTCPConnect.Enabled = true;
                btnTCPdiscon.Enabled = false;
            }
            catch (Exception)
            {
            }
        }

        //-------- Method ----------//
        private void GetAvailablePort()
        {
            cmbPortName.Items.Clear();
            String[] ports = SerialPort.GetPortNames();
            cmbPortName.Items.AddRange(ports);
        }

        public void TCPClientInitail()
        {
            Int32 port = Convert.ToInt32(txbPortNo.Text);
            String serverIP = txbIP.Text;
            IPAddress ip;
            bool IPvalid = IPAddress.TryParse(serverIP, out ip);
            if (IPvalid)
            {
                try
                {
                    client = new TcpClient(serverIP, port);
                    steam = client.GetStream();
                }
                catch (Exception)
                {
                }                
            }
            else
            {
                MessageBox.Show("IP address not correct");
            }
            
        }

        private void ControlMessage()
        {
            cmd[0] = Convert.ToByte(color.rgb.R);
            cmd[1] = Convert.ToByte(color.rgb.G);
            cmd[2] = Convert.ToByte(color.rgb.B);
            cmd[3] = Convert.ToByte(color.LEDPower);          
        }
    }
}

File: RGB_Control_VB/Form1.vb


Imports System.Net
Imports System.Net.Sockets
Imports Arduino_LED

Public Class Form1

    Dim color As MyRGB = New MyRGB
    Dim client As TcpClient = New TcpClient()
    Dim cmd() As Byte = {0, 0, 0, 255}
    Dim steam As NetworkStream

    Private Sub pnColor_Click(sender As Object, e As EventArgs) Handles pnColor.Click

        Dim colorResult As DialogResult = ColorDialog1.ShowDialog()
        If (colorResult = DialogResult.OK) Then
            color.setColor(ColorDialog1.Color)
            colorUpdate(color.rgb)

        End If

    End Sub

    Private Sub trbRGB_Scroll(sender As Object, e As EventArgs) Handles trbRed.Scroll, trbGreen.Scroll, trbBlue.Scroll
        color.setRED(trbRed.Value)
        color.setGreen(trbGreen.Value)
        color.setBLUE(trbBlue.Value)
        colorUpdate(color.rgb)
    End Sub

    Private Sub txbRGB_TextChanged(sender As Object, e As EventArgs) Handles txbRed.TextChanged, txbGrn.TextChanged, txbBlu.TextChanged

        If (txbRed.Text = "" Or txbGrn.Text = "" Or txbBlu.Text = "") Then
            If txbRed.Text = "" Then
                txbRed.Text = "0"
            End If
            If txbGrn.Text = "" Then
                txbGrn.Text = "0"
            End If
            If txbBlu.Text = "" Then
                txbBlu.Text = "0"
            End If
        ElseIf (Convert.ToInt32(txbRed.Text) > 255 Or
                Convert.ToInt32(txbGrn.Text) > 255 Or
                Convert.ToInt32(txbBlu.Text) > 255) Then
            If Convert.ToInt32(txbRed.Text) > 255 Then
                txbRed.Text = "255"
            End If
            If Convert.ToInt32(txbGrn.Text) > 255 Then
                txbGrn.Text = "255"
            End If
            If Convert.ToInt32(txbBlu.Text) > 255 Then
                txbBlu.Text = "255"
            End If
        End If

        color.setRED(Convert.ToInt32(txbRed.Text))
        color.setGreen(Convert.ToInt32(txbGrn.Text))
        color.setBLUE(Convert.ToInt32(txbBlu.Text))
        colorUpdate(color.rgb)

    End Sub

    Private Sub colorUpdate(rgb As Color)

        pnColor.BackColor = rgb

        trbRed.Value = rgb.R
        trbGreen.Value = rgb.G
        trbBlue.Value = rgb.B

        txbRed.Text = rgb.R
        txbGrn.Text = rgb.G
        txbBlu.Text = rgb.B

    End Sub

    Private Sub trbPow_Scroll(sender As Object, e As EventArgs) Handles trbPow.Scroll
        color.LEDPower = trbPow.Value
        txbPow.Text = trbPow.Value
    End Sub

    Private Sub txbPow_TextChanged(sender As Object, e As EventArgs) Handles txbPow.TextChanged

        If (txbPow.Text = "") Then
            txbPow.Text = "0"
        ElseIf (Convert.ToInt32(txbPow.Text) > 255) Then
            txbPow.Text = "255"
        End If
        color.LEDPower = Convert.ToInt32(txbPow.Text)
        trbPow.Value = color.LEDPower

    End Sub

    Private Sub rdbSerial_CheckedChanged(sender As Object, e As EventArgs) Handles rdbSerial.CheckedChanged
        gbSerial.Enabled = True
        cmbSerialPortNo.Enabled = True
        cmbSerialBaundRate.Enabled = True
        btnSerialConnect.Enabled = False
        btnSerialDiscon.Enabled = False

        cmbSerialPortNo.Items.Clear()
        Dim ports() As String = IO.Ports.SerialPort.GetPortNames
        cmbSerialPortNo.Items.AddRange(ports)

        gbTCP.Enabled = False
        If (client.Connected) Then
            client.Close()
        End If
        'disconnect TCP
    End Sub

    Private Sub SerialSetting_TextChanged(sender As Object, e As EventArgs) Handles cmbSerialPortNo.SelectedValueChanged, cmbSerialBaundRate.SelectedValueChanged
        If (cmbSerialPortNo.Text <> "" And cmbSerialBaundRate.Text <> "") Then
            SerialPort1.PortName = cmbSerialPortNo.Text
            SerialPort1.BaudRate = Convert.ToInt32(cmbSerialBaundRate.Text)
            btnSerialConnect.Enabled = True
        End If
    End Sub

    Private Sub btnSerialConnect_Click(sender As Object, e As EventArgs) Handles btnSerialConnect.Click
        Try
            SerialPort1.Open()
            btnSerialDiscon.Enabled = SerialPort1.IsOpen
            cmbSerialPortNo.Enabled = False
            cmbSerialBaundRate.Enabled = False
        Catch ex As Exception
            MessageBox.Show("Can't open port")
        End Try
    End Sub

    Private Sub btnSerialDiscon_Click(sender As Object, e As EventArgs) Handles btnSerialDiscon.Click
        SerialPort1.Close()
        btnSerialConnect.Enabled = False
        btnSerialDiscon.Enabled = False
        cmbSerialPortNo.Enabled = True
        cmbSerialBaundRate.Enabled = True
    End Sub

    Private Sub rdbTCP_CheckedChanged(sender As Object, e As EventArgs) Handles rdbTCP.CheckedChanged
        gbTCP.Enabled = True
        txbTcpIpAdd.Enabled = True
        txbTcpPortNo.Enabled = True
        btnSerialConnect.Enabled = False
        btnSerialDiscon.Enabled = False

        gbSerial.Enabled = False
        If SerialPort1.IsOpen Then
            SerialPort1.Close()
        End If
    End Sub

    Private Sub txbTcpIpAdd_TextChanged(sender As Object, e As EventArgs) Handles txbTcpIpAdd.TextChanged

        Dim ip As IPAddress = IPAddress.IPv6Loopback
        If (IPAddress.TryParse(txbTcpIpAdd.Text, address:=ip)) Then
            btnTcpConnect.Enabled = True
        Else
            btnTcpConnect.Enabled = False
        End If

    End Sub

    Private Sub btnTcpConnect_Click(sender As Object, e As EventArgs) Handles btnTcpConnect.Click

        Try
            client.Connect(txbTcpIpAdd.Text, Convert.ToInt32(txbTcpPortNo.Text))
            steam = client.GetStream()
            btnTcpDiscon.Enabled = True
            txbTcpIpAdd.Enabled = False
            txbTcpPortNo.Enabled = False
        Catch ex As Exception
            MessageBox.Show("Connection time out")
        End Try


    End Sub

    Private Sub btnTcpDiscon_Click(sender As Object, e As EventArgs) Handles btnTcpDiscon.Click

        client.Close()

        txbTcpIpAdd.Enabled = True
        txbTcpPortNo.Enabled = True
        btnTcpConnect.Enabled = False

    End Sub

    Private Sub btnColor_Click(sender As Object, e As EventArgs) Handles btnColor.Click

        cmd(0) = Convert.ToByte(color.rgb.R)
        cmd(1) = Convert.ToByte(color.rgb.G)
        cmd(2) = Convert.ToByte(color.rgb.B)
        cmd(3) = Convert.ToByte(color.LEDPower)

        If (SerialPort1.IsOpen) Then
            SerialPort1.Write(cmd, 0, 4)
        ElseIf (client.Connected) Then
            steam.Write(cmd, 0, 4)
        End If

    End Sub
End Class
